File: src/main/java/com/logicalgeekboy/logical_zoom/java_util/impl/MathUtil.java
package com.logicalgeekboy.logical_zoom.java_util.impl;

import java.math.BigDecimal;
import java.math.RoundingMode;

public class MathUtil {
    private static boolean DEBUG = true;

    private static void debug(String message) {
        if (DEBUG) {
            System.out.println("[MathUtil Debug] " + message);
        }
    }

    public static float findMiddleValue(float f, float f2, float f3) {
        float result = f < f2 ? f2 : Math.min(f, f3);
        if (DEBUG) {
            debug("findMiddleValue(" + f + ", " + f2 + ", " + f3 + ") = " + result);
        }
        return result;
    }

    public static float interpolate(float current, float target, float factor) {
        if (DEBUG) {
            debug("interpolate: current=" + current + ", target=" + target + ", factor=" + factor);
        }

        // Ensure factor is in valid range
        factor = findMiddleValue(factor, 0, 1);
        if (DEBUG && factor <= 0) {
            debug("WARNING: interpolation factor is <= 0! This will make no change to current value.");
        }

        float result = current + (target - current) * factor;
        debug("interpolation result: " + result + " (moved " + (factor * 100) + "% toward target)");
        return result;
    }

    public static float scaleAndRoundFloat(float n, int newScale) {
        if (newScale >= 0) {
            return BigDecimal.valueOf(n).setScale(newScale, RoundingMode.FLOOR).floatValue();
        }
        return -1;
    }
}

File: src/main/kotlin/com/dov/cm/util/MathUtil.kt
/*
masta codda - exotic
*/
package com.dov.cm.util

import java.math.BigDecimal
import java.math.RoundingMode
import kotlin.math.min

object MathUtil {
    fun findMiddleValue(f: Float, f2: Float, f3: Float): Float {
        return if (f < f2) f2 else min(f.toDouble(), f3.toDouble()).toFloat()
    }

    fun interpolate(n: Float, n2: Float, n3: Float): Float {
        return n - (n - n2) * findMiddleValue(n3, 0f, 1f)
    }

    fun scaleAndRoundFloat(n: Float, newScale: Int): Float {
        if (newScale >= 0) {
            return BigDecimal.valueOf(n.toDouble()).setScale(newScale, RoundingMode.FLOOR).toFloat()
        }
        return -1f
    }
}